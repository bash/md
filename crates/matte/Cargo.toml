[package]
name = "matte"
version = "0.1.0"
edition = "2021"

[dependencies]
anstyle.workspace = true
bat.workspace = true
fmtastic.workspace = true
hostname.workspace = true
matte-language-names = { path = "../matte-language-names" }
pulldown-cmark.workspace = true
smallvec.workspace = true
trait-set.workspace = true
unicode-linebreak-chunked.workspace = true
unicode-width.workspace = true
url.workspace = true

[dev-dependencies]
insta = { workspace = true, features = ["glob"] }
serde = { workspace = true, features = ["derive"] }
toml.workspace = true
