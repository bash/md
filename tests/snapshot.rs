use insta::{assert_snapshot, glob};
use matte::{render, supported_parser_options, Options};
use pulldown_cmark::{Event, MetadataBlockKind, Options as ParserOptions, Parser, Tag};
use serde::Deserialize;
use std::fs::read_to_string;

#[test]
fn test_snippets() {
    glob!("snippets/*.md", |path| {
        let markdown = read_to_string(path).unwrap();
        let rendered = render_to_string(&markdown);
        assert_snapshot!(rendered);
    })
}

fn render_to_string(input: &str) -> String {
    let parser = Parser::new_ext(input, supported_parser_options());
    let mut buffer = Vec::new();
    let snippet_options = read_snippet_options(input);
    let options = Options::plain_text(snippet_options.width.unwrap_or(120));
    render(parser, &mut buffer, options).unwrap();
    String::from_utf8(buffer).unwrap()
}

fn read_snippet_options(input: &str) -> SnippetOptions {
    toml::de::from_str(&extract_frontmatter(input)).unwrap()
}

fn extract_frontmatter(input: &str) -> String {
    let parser = Parser::new_ext(
        input,
        ParserOptions::ENABLE_PLUSES_DELIMITED_METADATA_BLOCKS,
    );
    let mut frontmatter = String::new();
    let mut inside_frontmatter = false;
    for event in parser {
        match event {
            Event::Start(Tag::MetadataBlock(MetadataBlockKind::PlusesStyle)) => {
                inside_frontmatter = true
            }
            Event::Text(text) if inside_frontmatter => frontmatter.push_str(&text),
            _ => break,
        }
    }
    frontmatter
}

#[derive(Deserialize)]
struct SnippetOptions {
    width: Option<u16>,
}
