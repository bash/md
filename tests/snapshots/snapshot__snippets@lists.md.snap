---
source: tests/snapshot.rs
expression: rendered
input_file: tests/snippets/lists.md
---
[1m[4m[32m[0m[1m[4m[32mLists[0m

[1m[32m1. [0m[1m[32mBlock Quote Followed by List[0m

‚îÉ[0m [0mBlock quote[0m

[1m‚Ä¢ [0mFoo[0m
[1m‚Ä¢ [0mBar[0m
[1m‚Ä¢ [0mBaz[0m

[1m[32m2. [0m[1m[32mOrdered List[0m

[1m1. [0mLorem ipsum dolor sit amet, [0m
[1m   [0mconsectetur adipiscing elit, sed do [0m
[1m   [0meiusmod tempor incididunt ut labore [0m
[1m   [0met dolore magna aliqua. Ut enim ad [0m
[1m   [0mminim veniam, quis nostrud [0m
[1m   [0mexercitation ullamco laboris nisi ut [0m
[1m   [0maliquip ex ea commodo consequat.[0m
[1m2. [0mDuis aute irure dolor in [0m
[1m   [0mreprehenderit in voluptate velit [0m
[1m   [0messe cillum dolore eu fugiat nulla [0m
[1m   [0mpariatur.[0m
[1m   [0m[1m1. [0mNested[0m
[1m   [0m[1m2. [0mThings[0m
[1m3. [0mExcepteur sint occaecat cupidatat [0m
[1m   [0mnon proident, sunt in culpa qui [0m
[1m   [0mofficia deserunt mollit anim id est [0m
[1m   [0mlaborum.[0m

[1m[32m3. [0m[1m[32mUnordered List[0m

[1m‚Ä¢ [0mFoo[0m
[1m‚Ä¢ [0mBar[0m
[1m‚Ä¢ [0mBaz[0m

[1m[32m4. [0m[1m[32mList Formatting[0m

[1m1. [0m[1mBold Text[0m[0m
[1m2. [0mCode Block[0m
[1m   [0mconsole[35m.[0m[36mlog[0m([32m'[0m[32mhello world[0m[32m'[0m)
[1m   [0mwith text after code block[0m
[1m3. [0mSublist with text before[0m
[1m   [0m[1m‚ó¶ [0mfoo[0m
[1m   [0m[1m‚ó¶ [0mbar[0m
[1m   [0m[1m‚ó¶ [0mbaz and text after[0m

[1m[32m5. [0m[1m[32mQuoted List[0m

‚îÉ[0m [0m[1m1. [0mWow[0m
‚îÉ[0m [0m[1m2. [0mWhat[0m
‚îÉ[0m [0m[1m3. [0mA[0m
‚îÉ[0m [0m[1m4. [0mQuote[0m

[1m[32m6. [0m[1m[32mDeeply Nested List[0m

[1m1. [0mWow[0m
[1m   [0m[1m1. [0mMuch[0m
[1m   [0m[1m   [0m[1m1. [0mNesting[0m
[1m   [0m[1m   [0m[1m   [0m[1m1. [0mSuch[0m
[1m   [0m[1m   [0m[1m   [0m[1m   [0m[1m1. [0mDeep[0m
[1m   [0m[1m   [0m[1m   [0m[1m   [0m[1m   [0m[1m1. [0mList[0m

[1m‚Ä¢ [0mWow[0m
[1m  [0m[1m‚ó¶ [0mMuch[0m
[1m  [0m[1m  [0m[1m‚ñ™ [0mNesting[0m
[1m  [0m[1m  [0m[1m  [0m[1m‚Ä£ [0mSuch[0m
[1m  [0m[1m  [0m[1m  [0m[1m  [0m[1m‚Ä¢ [0mDeep[0m
[1m  [0m[1m  [0m[1m  [0m[1m  [0m[1m  [0m[1m‚ó¶ [0mList[0m

[1m[32m7. [0m[1m[32mTasks[0m

‚òê [0mbuy groceries[0m
‚òí [0mclean kitchen[0m
‚òí [0mfix bathroom lights[0m
